src/internal/communication.ts

```diff
@@ -0,0 +1,337 @@
+  public static childOrigin: string;
+}
+
+class CommunicationPrivate {
+  public static parentMessageQueue: MessageRequest[] = [];
+  public static childMessageQueue: MessageRequest[] = [];
+  public static nextMessageId: number = 0;
+  public static callbacks: {
+    [id: number]: Function;
+  } = {};
+  public static messageListener: Function;
+}
+
+export function initializeCommunication(callback: Function, validMessageOrigins: string[] | undefined): void {
+  // Listen for messages post to our window
+  CommunicationPrivate.messageListener = (evt: DOMMessageEvent): void => processMessage(evt);
+
+  // If we are in an iframe, our parent window is the one hosting us (i.e., window.parent); otherwise,
+  // it's the window that opened us (i.e., window.opener)
+  Communication.currentWindow = Communication.currentWindow || window;
+  Communication.parentWindow =
+    Communication.currentWindow.parent !== Communication.currentWindow.self
+      ? Communication.currentWindow.parent
+      : Communication.currentWindow.opener;
+
+  // Listen to messages from the parent or child frame.
+  // Frameless windows will only receive this event from child frames and if validMessageOrigins is passed.
+  if (Communication.parentWindow || validMessageOrigins) {
+    Communication.currentWindow.addEventListener('message', CommunicationPrivate.messageListener, false);
+  }
+
+  if (!Communication.parentWindow) {
+    GlobalVars.isFramelessWindow = true;
+    // @ts-ignore: window as ExtendedWindow
+    (window as ExtendedWindow).onNativeMessage = handleParentMessage;
+  }
+
+  try {
+    // Send the initialized message to any origin, because at this point we most likely don't know the origin
+    // of the parent window, and this message contains no data that could pose a security risk.
+    Communication.parentOrigin = '*';
+    sendMessageToParent('initialize', [version], callback);
+  } finally {
+    Communication.parentOrigin = null;
+  }
+}
+
+export function uninitializeCommunication(): void {
+  Communication.currentWindow.removeEventListener('message', CommunicationPrivate.messageListener, false);
+
+  Communication.parentWindow = null;
+  Communication.parentOrigin = null;
+  Communication.childWindow = null;
+  Communication.childOrigin = null;
+  CommunicationPrivate.parentMessageQueue = [];
+  CommunicationPrivate.childMessageQueue = [];
+  CommunicationPrivate.nextMessageId = 0;
+  CommunicationPrivate.callbacks = {};
+}
+
+/**
+ * Send a message to parent. Uses nativeInterface on mobile to communicate with parent context
+ */
+export function sendMessageToParent(actionName: string, callback?: Function): void;
+/**
+ * Send a message to parent. Uses nativeInterface on mobile to communicate with parent context
+ */
+export function sendMessageToParent(actionName: string, args: any[], callback?: Function): void;
+export function sendMessageToParent(actionName: string, argsOrCallback?: any[] | Function, callback?: Function): void {
```

Comment:
why do we define this function 3 times?

Reply:
This will create better documentation. The first two are declarations of possible interfaces (with or without args), the last one the implementation.
https://www.typescriptlang.org/docs/handbook/functions.html#overloads